src/simd/allocator.h

```diff
@@ -0,0 +1,165 @@
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/** @file allocator.h
+ *
+ * Provide a custom memory allocator for SIMD.
+ *
+ * This allocator always returns memory that is suitably aligned to be loaded
+ * efficiently by a Register object.
+ */
+
+#ifndef __QUAD_SIMD_SIMD_ALLOCATOR_H__
+#define __QUAD_SIMD_SIMD_ALLOCATOR_H__
+
+#include <cassert>
+#include <cstdint>
+#include <limits>
+
+#include "simd/definitions.h"
+
+namespace quadiron {
+namespace simd {
+
+/// Check that the given address is properly aligned.
+template <typename T>
+inline bool addr_is_aligned(const T* addr)
+{
+    // Without SIMD, there is no specific alignment constraint.
+    if (INSTRUCTION_SET == InstructionSet::NONE) {
+        return true;
+    }
+    const std::uintptr_t address = reinterpret_cast<std::uintptr_t>(addr);
+    return (address & (ALIGNMENT - 1)) == 0;
+}
+
+/** Custom allocator to take advantage of SIMD processing.
+ *
+ * This allocator always return memory that is suitably aligned for the current
+ * SIMD instruction set. Thanks to this property, you can safely use the aligned
+ * load from the Register class in order to increase performance.
+ */
+template <typename T>
+class AlignedAllocator {
+  public:
+    using value_type = T;
+
+    AlignedAllocator() noexcept {}
+    // No state, => nothing to copy.
+    template <class U>
+    AlignedAllocator(AlignedAllocator<U> const& other) noexcept
+    {
+    }
+
+    value_type* allocate(std::size_t count)
+    {
+        // Guard against overflow!
+        if (count > max_size()) {
+            throw std::bad_alloc();
+        }
+
+        // No SIMD: default allocator is good enough!
+        if (INSTRUCTION_SET == InstructionSet::NONE) {
+            return static_cast<value_type*>(
```

Comment:
Why don't we use `new T[count]`?

Reply:
There is a difference between those two.
- the form I used is the "raw" `operator new` (aka the global `operator new`), which behaves like `malloc` in C: it returns a block of uninitialized storage (no object has been constructed within that storage, thus it has no dynamic type).
- the form you suggest is the most common and combines two steps: first the memory is allocated (using the "raw" operator) and then it construct an object of type T (hence this piece of memory has a dynamic type T).

Now, does it matter? After all, in our very specific case we mainly allocate native integer types, which doesn't have a constructor so it should make no difference, right?
In fact it does matter, because of strict aliasing rules: in the previous version of our allocation function (which uses `new uint8_t[len]`) we weren't allocation a piece of memory (like `malloc` does in C), we were creating an array of `uint8_t` from then on we were subject to the strict aliasing rule (the line `T* data = reinterpret_cast<T*>(ptr);` was breaking this rule and thus was an undefined behavior).

The new version, that uses the "raw" `operator new`, return untyped memory (no object are constructed) and thus isn't subject to this rule (because no dynamic type was assigned) until we construct an object either by using placement new (which is another version of the `operator new`), or by calling the method `construct` of the allocator (which usually relies on placement new).

If you wan a quick overview of the different facet of `new` you can read [this blog post](https://eli.thegreenplace.net/2011/02/17/the-many-faces-of-operator-new-in-c).

See also [this](https://stackoverflow.com/a/2697929) and [this](https://stackoverflow.com/a/36021642) which may be more readable than the C++ standard.

Reply:
Okay. So does `operator new` allocates a contiguous piece of memory?

Reply:
Yes, of course `new` return contiguous memory.
Otherwise, how the language would work (no array, no std::vector, …)?

Reply:
yes, okay :)